lib/landrush/win_network_config.rb

```diff
@@ -0,0 +1,159 @@
+    def self.which(cmd)
+      exts = ENV['PATHEXT'] ? ENV['PATHEXT'].split(';') : ['']
+      ENV['PATH'].split(File::PATH_SEPARATOR).each do |path|
+        exts.each do |ext|
+          exe = File.join(path, "#{cmd}#{ext}")
+          return exe if File.executable?(exe) && !File.directory?(exe)
+        end
+      end
+      nil
+    end
+
+    # If this registry query succeeds we assume we have Admin rights
+    # http://stackoverflow.com/questions/8268154/run-ruby-script-in-elevated-mode/27954953
+    def self.admin_mode?
+      (`reg query HKU\\S-1-5-19 2>&1` =~ /ERROR/).nil?
+    end
+
+    private
+
+    # Given a network name (as displayed on 'Control Panel\Network and Internet\Network Connections'),
+    # determines the GUID of this network interface using 'netsh'.
+    #
+    # To make this work the "Wired Autoconfig" service must be started (go figure).
+    #
+    # Output of netsh command which is being processed:
+    #
+    # There are 4 interfaces on the system:
+    #
+    # Name             : Ethernet
+    # Description      : Intel(R) Ethernet Connection (3) I218-LM
+    # GUID             : fd9270f6-aff6-4f24-bc4a-1f90c032d5c3
+    # Physical Address : 50-7B-9D-AB-25-1D
+    # \n\n
+    # ...
+    def get_guid(network_name)
+      cmd_out = `netsh lan show interfaces`
+      interface_details = cmd_out.split(/\n\n/).select { |settings| settings.match(/#{Regexp.quote(network_name)}/m) }
+      interface_details[0].split(/\n/)[2].match(/.*:(.*)/).captures[0].strip
+    end
+
+    # Given an IP determines the network name, if any. Uses netsh which generates output like this:
+    #
+    # ...
+    # \n\n
+    # Configuration for interface "Ethernet 2"
+    #    DHCP enabled:                         Yes
+    #    IP Address:                           10.10.10.1
+    #    Subnet Prefix:                        10.10.10.0/24 (mask 255.255.255.0)
+    #    InterfaceMetric:                      10
+    #    DNS servers configured through DHCP:  None
+    #    Register with which suffix:           Primary only
+    #    WINS servers configured through DHCP: None
+    # \n\n
+    # ...
+    def get_network_name(ip)
+      cmd_out = `netsh interface ip show config`
+      network_details = cmd_out.split(/\n\n/).select do |settings|
+        lines = settings.split(/\n/).reject(&:empty?)
+        subnet = lines[3]
+        next false unless subnet =~ /Subnet Prefix/
+
+        mask = IPAddr.new(subnet.match(%r{.* (\d{1,3}\.\d{1,3}\.\d{1,3}.\d{1,3}/\d{1,3}).*}).captures[0])
+        address = IPAddr.new(ip)
+
+        mask.include?(address)
+      end
+      return nil if network_details[0].nil?
+      network_details[0].split(/\n/)[0].match(/Configuration for interface "(.*)"/).captures[0].strip
+    end
+
+    # Makes sure that we have admin privileges and if nor starts a new shell with the required
+    # privileges
+    def ensure_admin_privileges(file, *args)
+      unless self.class.admin_mode?
+        require 'win32ole'
```

Comment:
but here is is required. Trying to require win32ole on Linux will fail and throw LoadErrors. In this case we need to defer the loading until we know that we are on Windows. In the case we had the other day, it was a require for a plain gem afair.